ecpy_hqc_legacy/instruments/drivers/visa/cryomagnetics_cs4.py

```diff
@@ -44,54 +56,81 @@ def __init__(self, connection_info, caching_allowed=True,
 
-    @secure_communication()
-    def make_ready(self):
-        """Setup the correct unit and range.
-
-        """
+        # Setup the correct unit and range.
         self.write('UNITS T')
-        self.write('RANGE 0 100')
+        self.write('RANGE 0 100;')  # HINT the CG4 requires the ;
         # we'll only use the command sweep up (ie to upper limit)
         # however upper limit can't be lower than lower limit for
         # some sources : G4 for example
         # set lower limit to lowest value
         self.write('LLIM -7')
 
-    def go_to_field(self, value, rate, auto_stop_heater=True,
-                    post_switch_wait=30):
-        """Ramp up the field to the specified value.
+    def read_output_field(self):
+        """Read the current value of the output field.
 
         """
-        self.field_sweep_rate = rate
+        return float(self.ask('IOUT?').strip(' T'))
 
-        if abs(self.persistent_field - value) >= OUT_FLUC:
+    def read_persistent_field(self):
+        """Read the current value of the persistent field.
 
-            if self.heater_state == 'Off':
-                self.target_field = self.persistent_field
-                self.heater_state = 'On'
-                sleep(1)
-
-            self.target_field = value
-
-        if auto_stop_heater:
-            self.heater_state = 'Off'
-            sleep(post_switch_wait)
-            self.activity = 'To zero'
-            wait = 60 * abs(self.target_field) / self.field_sweep_rate
-            sleep(wait)
-            niter = 0
-            while abs(self.target_field) >= OUT_FLUC:
-                sleep(5)
-                niter += 1
-                if niter > MAXITER:
-                    raise InstrIOError(cleandoc('''CS4 didn't set the
-                        field to zero after {} sec'''.format(5 * MAXITER)))
+        """
+        return float(self.ask('IMAG?').strip(' T'))
+
+    def is_target_reached(self):
+        """Check whether the target field has been reached.
+
+        """
+        return (abs(self.read_output_field() - self.target_field) <
+                self.output_fluctuations)
+
+    def sweep_to_persistent_field(self):
+        """Convenience function ramping the field to the persistent.
+
+        Once the value is reached one can safely turn on the switch heater.
+
+        """
+        return self.sweep_to_field(self.read_persistent_field())
+
+    def sweep_to_field(self, value, rate=None):
+        """Convenience function to ramp up the field to the specified value.
+
+        """
+        # Set rate. Always the fast sweep rate if the switch heater is off.
+        if rate is not None:
+            self.field_sweep_rate = rate
+        rate = (self.field_sweep_rate if self.heater_state == 'Off' else
+                self.fast_sweep_rate)
+
+        # Start ramping.
+        self.target_field = value
+        self.activity = 'To set point'
+
+        # Create job.
+        span = abs(self.out_field - value)
```

Comment:
out_field does not exist anymore.
the span is : 
- persistent_field - value if switch heater if on
- output_field - persistent_field if switch heater is off

Reply:
I fixed the reference to out_field. However, no matter the state of the switch heater (this does not depend on it), the span is output_field (as returned by read_output_field) - value. I need to figure out how to do that cleanly for the IPS.